### ISSUE TITLE
Compile flags from profile not passed to CMake is using targets

### ISSUE THREAD

**Issue by mropert** [conan-io/conan#1111]
I've discovered a regression in 0.20 if you use the CMake generator used with the "targets" methods.

Today the compiler flags from the profile (for example, the ```-m32``` or ```-m64``` derived from the ```arch``` setting) are passed to CMake as conan env variables (```CONAN_CXX_FLAGS```).

When not using targets, those flags are directly set as CMake's in the ```conan_global_flags``` macro:
```    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CONAN_CXX_FLAGS}")```

In the case where Conan build with CMake targets (passing ```TARGETS``` to ```conan_basic_setup()```), that macro is never called and the flags never impact the build unless added by hand by the recipe maintainer in his ```CMakeLists.txt```.

On Solaris it breaks most of my recipes because the compiler defaults to 32 bit binaries even when Conan's profile is set to 64 bits.

**Comment by mropert** [comment]
It might not be a regression after, just tested with 0.19.3 and same issue.

**Comment by mropert** [comment]
As a workaround I've added those lines at the top of my ```CMakeLists.txt``` after calling ```conan_basic_setups(TARGETS)```:
```
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CONAN_CXX_FLAGS}")
set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${CONAN_C_FLAGS}")
set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} ${CONAN_SHARED_LINKER_FLAGS}")
set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${CONAN_EXE_LINKER_FLAGS}")
```

But I feel it should be automated / handled by Conan

**Comment by lasote** [comment]
Could you post your conanbuildinfo.cmake?

**Comment by mropert** [comment]
You can check it out there: http://pastebin.com/cLAfXust

**Comment by lasote** [comment]
Thanks! 
Each target only receives the inherited flags from the `cpp_info` objects, and it has kind of sense.
I'm not sure if conan:
    - should set the global flags (it doesn't look well)
    - should automatically apply to every target the global flags.
    - should provide a global target interface to be applied too.

Maybe @memsharded can help us with the best approach of telling us if the current behavior is intended for some reason. 

Anyway It doesn't look hard to solve and you can (as you are doing) set the flags to your targets or globally, so it's not a blocking issue, right?

**Comment by mropert** [comment]
It's not blocking since I have a workaround :)
Still, I think it would be interesting to really reflect on how we expect the ```TARGETS``` to be used.
A good exercise, I think, is to do like I did and try to migrate a project from globals to targets and see what happens.

**Comment by memsharded** [comment]
Quick question. The ``-m32`` , ``-m64`` flags, are defined in your profile? Or ar they set by the ``CMake`` conan helper in the recipe ``build()`` method. I'd say the second one.

Yes, the design rationale with TARGETS is that no global settings should be done, as the purpose of TARGETS is precisely to only get the flags that really affect you.

The problem with 
```cmake
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CONAN_CXX_FLAGS}")
```
is that it will activate all the flags, including the dependencies ones, which are already defined in the TARGETS.

So two things are needed:
- First, introduce a modification in the generated ``conanbuildinfo.cmake``, so the provided flags are stored separately.
- I would do one of these two:
   - Set the global cmake flags. Not that dirty, they are quite global settings (for the package). And it will be transparent and easy for users
   - Define another "PKG_TARGET" that will contain the flags. Consumers will have to link with this other target for every of their targets.

I would say I prefer the first one.

**Comment by lasote** [comment]
I've thought about it and yes, I agree, setting the global CMake flags is not bad, they are global flags.

**Comment by mropert** [comment]
I agree too, I think the set to the ```${CMAKE_CXX_FLAGS}``` (and co) should done regardless of the setup mode (targets or non targets).

**Comment by memsharded** [comment]
Finally, the global approach has been implemented and released in 0.21. Please @mropert try to remove your workaround and test if it works now. Thanks!
